modules/messages/server/services/message-to-influx.server.service.js

```diff
@@ -0,0 +1,192 @@
+ * @param {influxCallback} callback - a callback that handles the response
+ */
+module.exports.save = function (message, callback) {
+  async.waterfall([
+    function (done) {
+      return done(null, message);
+    },
+    module.exports.process,
+    module.exports.send
+  ], function (err, response) {
+    if (err) console.error(err);
+    if (typeof callback === 'function') {
+      return callback(err, response);
+    }
+  });
+};
+
+/**
+ * this function gets some info about the message sent and then calls the callback
+ * with field and tag keys and values as they're expected in database
+ * @param {object} message - a message object as returned by mongoDB
+ * @param {processMessageCallback} callback
+ */
+module.exports.process = function (message, callback) {
+  // some variables used later, filled inside waterfall
+  var isFirstMessage,
+      isFirstReply;
+
+  // fixing some strange filling of message data
+  // (userFrom is not id but user object)
+  var userFrom = message.userFrom._id
+    ? message.userFrom._id
+    : message.userFrom;
+  var userTo = message.userTo._id
+    ? message.userTo._id
+    : message.userTo;
+
+  async.waterfall([
+    function readFirstMessage(done) {
+      // find the oldest message of the thread
+      return Message.findOne({
+        $or: [
+          {
+            userTo: userTo,
+            userFrom: userFrom
+          },
+          {
+            userTo: userFrom,
+            userFrom: userTo
+          }
+        ]
+      })
+      .sort({ created: 1 })
+      .exec(done);
+    },
+
+    function readFirstReply(firstMessage, done) {
+      // if no message was found, throw error (there is always the first message
+      // already (at least the one just saved))
+      if (!firstMessage) {
+        var err = new Error('first message not found, but should have been already saved');
+        done(err);
+      }
+
+
+      // is the new message the first message of the thread?
+      isFirstMessage = String(firstMessage._id) === String(message._id);
+
+      // can the message be the actual first reply?
+      // - is it not the firstMessage?
+      // is the sender and receiver in different order than in the firstMessage?
+      var canBeTheFirstReply = !isFirstMessage && String(firstMessage.userTo) === String(userFrom);
+      // if this can be the oldest reply, find the oldest reply of the thread
+      if (canBeTheFirstReply) {
+        return Message.findOne({
+          userTo: firstMessage.userFrom,
+          userFrom: firstMessage.userTo
+        })
+        .sort({ created: 1 })
+        .exec(function (err, firstReply) {
+          return done(err, firstMessage, firstReply);
+        });
+      } else {
+        return done(null, firstMessage, null);
+      }
+    },
+
+    function prepareData(firstMessage, firstReply, done) {
+      // is the new message the first reply of the thread?
+      isFirstReply = !!(firstReply && String(firstReply._id) === String(message._id));
```

Comment:
thanks for pointing out that `equals`. it will be useful. Here i'll leave strings because i don't understand it, yet.

```
null && false // returns null
```
i changed `!!` to `Boolean()`, but prefer to keep it (although it may make no difference). Without it possible outcomes would be `null`, `false`, `true`.

that check is kind of crappy here. but could be enough since we expect only null or object with _id. i don't know.